### PR TITLE
Implement Self Signed Root in SSZ

### PR DESCRIPTION
**Pull request by nisdas**
Part of #2307

### CODE REVIEW

File: shared/ssz/hash_test.go [hunk 1]
```diff
@@ -161,6 +169,22 @@ var merkleHashTests = []merkleHashTest{
...
+	{val: &pb.VoluntaryExit{Signature: []byte("SigningExit")}, output: "C940F7E42FF91F069939264514E6F88F690615FF452B76BECFFD5D36063EAAE4"},
+	{val: pb.BeaconBlockHeader{Slot: 20, Signature: []byte{'A', 'B'}}, output: "15DE30A770592FAD0AC0AE262DBEF783AA747DA5103275F651E41DF1E8CC653F"},
+	{val: pb.BeaconBlockHeader{Slot: 10, Signature: []byte("TESTING")}, output: "85AA0ABF2B8CEC5DB303320328B691FA125B3A4D3106365A29EAE44C9E60BF07"},
+	{val: pb.IndexedAttestation{Signature: []byte("SigningAttestation")}, output: "F00354BFA20B0CFBBE623A1AE36D9CE0D29FF65C3D69BF133C264B0707579E3B"},
+	{val: pb.VoluntaryExit{Signature: []byte("SigningExit")}, output: "C940F7E42FF91F069939264514E6F88F690615FF452B76BECFFD5D36063EAAE4"},
+	{val: pb.Deposit{Index: 0}, error: "field name is invalid wanted Signature but got"},
+	{val: 2, error: "given object is neither a struct or a pointer"},
+	{val: []byte{'a'}, error: "given object is neither a struct or a pointer"},
+	{val: nil, error: "given object is neither a struct or a pointer"},
+	{val: (*[]uint8)(nil), error: "nil pointer given"},
```
Comment (shayzluf) [inline shared/ssz/hash_test.go]: missing test for `signature` field name with small letters

File: shared/ssz/hash_test.go [hunk 2]
```diff
@@ -161,6 +169,22 @@ var merkleHashTests = []merkleHashTest{
 	}, output: "55DC6699E7B5713DD9102224C302996F931836C6DAE9A4EC6AB49C966F394685"},
 }
 
+var signatureRootTests = []signatureRootTest{
+	{val: &pb.BeaconBlockHeader{Slot: 20, Signature: []byte{'A', 'B'}}, output: "15DE30A770592FAD0AC0AE262DBEF783AA747DA5103275F651E41DF1E8CC653F"},
+	{val: &pb.BeaconBlockHeader{Slot: 10, Signature: []byte("TESTING")}, output: "85AA0ABF2B8CEC5DB303320328B691FA125B3A4D3106365A29EAE44C9E60BF07"},
+	{val: &pb.IndexedAttestation{Signature: []byte("SigningAttestation")}, output: "F00354BFA20B0CFBBE623A1AE36D9CE0D29FF65C3D69BF133C264B0707579E3B"},
+	{val: &pb.VoluntaryExit{Signature: []byte("SigningExit")}, output: "C940F7E42FF91F069939264514E6F88F690615FF452B76BECFFD5D36063EAAE4"},
+	{val: pb.BeaconBlockHeader{Slot: 20, Signature: []byte{'A', 'B'}}, output: "15DE30A770592FAD0AC0AE262DBEF783AA747DA5103275F651E41DF1E8CC653F"},
+	{val: pb.BeaconBlockHeader{Slot: 10, Signature: []byte("TESTING")}, output: "85AA0ABF2B8CEC5DB303320328B691FA125B3A4D3106365A29EAE44C9E60BF07"},
```
Comment (shayzluf) [inline shared/ssz/hash_test.go]: missing test for more then two fields

File: shared/ssz/hash_test.go [hunk 3]
```diff
@@ -161,6 +169,22 @@ var merkleHashTests = []merkleHashTest{
 	}, output: "55DC6699E7B5713DD9102224C302996F931836C6DAE9A4EC6AB49C966F394685"},
 }
 
+var signatureRootTests = []signatureRootTest{
+	{val: &pb.BeaconBlockHeader{Slot: 20, Signature: []byte{'A', 'B'}}, output: "15DE30A770592FAD0AC0AE262DBEF783AA747DA5103275F651E41DF1E8CC653F"},
+	{val: &pb.BeaconBlockHeader{Slot: 10, Signature: []byte("TESTING")}, output: "85AA0ABF2B8CEC5DB303320328B691FA125B3A4D3106365A29EAE44C9E60BF07"},
+	{val: &pb.IndexedAttestation{Signature: []byte("SigningAttestation")}, output: "F00354BFA20B0CFBBE623A1AE36D9CE0D29FF65C3D69BF133C264B0707579E3B"},
+	{val: &pb.VoluntaryExit{Signature: []byte("SigningExit")}, output: "C940F7E42FF91F069939264514E6F88F690615FF452B76BECFFD5D36063EAAE4"},
+	{val: pb.BeaconBlockHeader{Slot: 20, Signature: []byte{'A', 'B'}}, output: "15DE30A770592FAD0AC0AE262DBEF783AA747DA5103275F651E41DF1E8CC653F"},
```
Comment (shayzluf) [inline shared/ssz/hash_test.go]: missing test for signature type of uint8

File: shared/ssz/hash_test.go [hunk 4]
```diff
@@ -161,6 +169,22 @@ var merkleHashTests = []merkleHashTest{
...
+	{val: pb.BeaconBlockHeader{Slot: 10, Signature: []byte("TESTING")}, output: "85AA0ABF2B8CEC5DB303320328B691FA125B3A4D3106365A29EAE44C9E60BF07"},
+	{val: pb.IndexedAttestation{Signature: []byte("SigningAttestation")}, output: "F00354BFA20B0CFBBE623A1AE36D9CE0D29FF65C3D69BF133C264B0707579E3B"},
+	{val: pb.VoluntaryExit{Signature: []byte("SigningExit")}, output: "C940F7E42FF91F069939264514E6F88F690615FF452B76BECFFD5D36063EAAE4"},
```
Comment (shayzluf) [inline shared/ssz/hash_test.go]: missing test for signature as a struct

File: shared/ssz/hash_test.go (hunk 1 above)
Comment (nisdas) [reply shared/ssz/hash_test.go]: thinking about this more, none of our core data structures have a 'signature' field , they all have the first letter capitalized. Maybe we should just remove the requirement. What do you think ?

File: shared/ssz/hash_test.go (hunk 3 above)
Comment (shayzluf) [reply shared/ssz/hash_test.go]: nvm no need to support

File: shared/ssz/hash_test.go (hunk 1 above)
Comment (shayzluf) [reply shared/ssz/hash_test.go]: what does the spec say. maybe we will get messages from other clients that will not be capitalized

**Pull request merged**